### PR TITLE
(SIMP-7096) Experimental Firewalld Support

### PR DESCRIPTION
**Pull request by trevor-vaughan**
- Added support for voxpupuli/firewalld
- Added an `iptables::firewalld_shim` class for configuring firewalld in
  accordance with the current iptables configuration.
- Added auto-translation capabilities to convert current calls to the
  `iptables::rule` defined type to firewalld rulesets.
- No work has been done to convert the advanced capabilities such as
  scanblock.
- Will automatically use firewalld in EL 8
- Allows optional enabling of firewalld in EL 7 by either setting
  `iptables::use_firewalld` to `true`
- IPv6 is not currently supported due to needing to move the auto-detection
  logic to a higher level. Preferrably, this will go into the upstream
  firewalld module.

SIMP-7098 #close

### CODE REVIEW

File: manifests/listen/all.pp [hunk 1]
```diff
@@ -68,11 +68,27 @@
...
-    first    => $first,
-    absolute => $absolute,
-    order    => $order,
-    apply_to => $apply_to,
-    content  => template("${module_name}/allow_all_services.erb")
+  include 'iptables'
+
+  if $iptables::use_firewalld {
+    $metadata = {
+      'dports'       => $dports,
```
Comment (lnemsick-simp) [inline manifests/listen/all.pp]: $dports is not defined in this manifest.

File: manifests/rule.pp [hunk 2]
```diff
@@ -85,13 +85,95 @@
...
+        require => Service['firewalld']
+      }
+    }
+    else {
+      # Create a unique ipset based on every content collection
+      #
+      # This is done so that we do not end up with a million ipsets for every call
+      #
+      # The length is limited due to apparent limitations in the ipset name
+      $_ipset_name = join(['simp',sha1(join(sort(unique($_trusted_nets_a)),''))], '_')[0,31]
```
Comment (lnemsick-simp) [inline manifests/rule.pp]: $_trusted_nets_a is not available in this manifest

File: manifests/listen/all.pp (hunk 1 above)
Comment (trevor-vaughan) [reply manifests/listen/all.pp]: Good catch, thanks.

Comment (trevor-vaughan) [reply manifests/listen/all.pp]: Re-reading, I just realized that I didn't handle this case properly. It's not used by SIMP by default so was missed.

**Comment by trevor-vaughan** [comment]
I started adding more advanced acceptance tests and ran into some issues with firewalld itself. Marking WIP until I can figure it out.

File: manifests/rule.pp [hunk 3]
```diff
@@ -26,6 +26,22 @@
 # @param content
 #   The **exact** content of the rule that should be added
 #
+#   * If this module is run in ``firewalld`` mode, this needs to contain a JSON
```
Comment (lnemsick-simp) [inline manifests/rule.pp]: This looks like a breaking change to me.  I think a new defined type should have been added instead iptables::firewalld_rule

File: manifests/firewalld_shim.pp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+#   * Set to ``true`` to re-enable this capability.
+#
+# @param lockdown
+#   Set ``firewalld`` in ``lockdown`` mode which disallows manipulation by
+#   applications.
+#
+#   * This makes sense to do by default since puppet is meant to be
+#     authoritative on the system.
+#
+# @param default_zone
```
Comment (lnemsick-simp) [inline manifests/firewalld_shim.pp]: Need to note that if this not 'simp', a 'simp' zone will still be created and all rules from this module will be applied to the 'simp' zone.

File: manifests/init.pp [hunk 5]
```diff
@@ -80,7 +83,8 @@
 # @author https://github.com/simp/pupmod-simp-iptables/graphs/contributors
 #
 class iptables (
-  Variant[Enum['ignore'],Boolean] $enable                     = simplib::lookup('simp_options::firewall', { 'default_value' => true }),
+  Variant[Enum['ignore','firewalld'],Boolean] $enable         = simplib::lookup('simp_options::firewall', { 'default_value' => true }),
```
[the remark this answers is not in the document]
Comment (trevor-vaughan) [reply manifests/init.pp]: Yes, once this goes through in some form.

File: manifests/init.pp [hunk 6]
```diff
@@ -95,47 +99,66 @@
...
-    if $ipv6 and $facts['ipv6_enabled'] {
-      file { '/etc/sysconfig/ip6tables':
-        owner => 'root',
-        group => 'root',
-        mode  => '0640'
+      file { '/etc/sysconfig/iptables':
+        owner   => 'root',
+        group   => 'root',
+        mode    => '0640',
+        require => Class['iptables::install']
```
[the remark this answers is not in the document]
Comment (trevor-vaughan) [reply manifests/init.pp]: If we want it to force order binding, yes because it's in the same class.

File: manifests/rule.pp (hunk 3 above)
Comment (trevor-vaughan) [reply manifests/rule.pp]: That seems reasonable.

File: manifests/firewalld_shim.pp (hunk 4 above)
Comment (trevor-vaughan) [reply manifests/firewalld_shim.pp]: Good idea.

File: manifests/rule.pp (hunk 3 above)
Comment (trevor-vaughan) [reply manifests/rule.pp]: @lnemsick-simp Your suggestion has turned out to be 100% better than the original implementation in all ways. Thanks!

**Comment by trevor-vaughan** [comment]
⏳ https://gitlab.com/simp/pupmod-simp-iptables/pipelines/98732851

**Comment by trevor-vaughan** [comment]
⏳ https://gitlab.com/simp/pupmod-simp-iptables/pipelines/98733732

**Comment by trevor-vaughan** [comment]
⏳ https://gitlab.com/simp/pupmod-simp-iptables/pipelines/98771603

**Comment by trevor-vaughan** [comment]
:hourglass_flowing_sand: https://gitlab.com/simp/pupmod-simp-iptables/pipelines/100142823

**Comment by trevor-vaughan** [comment]
GitLab runners are having issues. Will restart the test when possible.

**Comment by trevor-vaughan** [comment]
Runners are back online and the tests passed

File: spec/acceptance/nodesets/rhel8.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      - server
+      - default
+      - master
+      - simp_server
+    platform:   el-8-x86_64
+    box:        generic/rhel8
+    hypervisor: <%= hypervisor %>
+    yum_repos:
+      epel:
+        mirrorlist: 'https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=$basearch'
```
Comment (lnemsick-simp) [inline spec/acceptance/nodesets/rhel8.yml]: epel-7 ? Should this be epel-8

File: manifests/listen/all.pp [hunk 8]
```diff
@@ -62,17 +62,29 @@
...
-  iptables_rule { "all_${name}":
-    first    => $first,
-    absolute => $absolute,
-    order    => $order,
-    apply_to => $apply_to,
-    content  => template("${module_name}/allow_all_services.erb")
+  include 'iptables'
+
+  if $iptables::use_firewalld {
+    iptables::firewalld::rule { "all_${name}":
```
Comment (lnemsick-simp) [inline manifests/listen/all.pp]: $first does not apply?

File: manifests/listen/icmp.pp [hunk 9]
```diff
@@ -71,18 +71,31 @@
...
-    first    => $first,
-    absolute => $absolute,
-    order    => $order,
-    apply_to => $apply_to,
-    content  => template("${module_name}/allow_icmp_services.erb")
+  include 'iptables'
+
+  if $iptables::use_firewalld {
+    iptables::firewalld::rule { "icmp_${name}":
+      trusted_nets => $trusted_nets,
```
Comment (lnemsick-simp) [inline manifests/listen/icmp.pp]: $first and $absolute do not apply?

File: manifests/listen/tcp_stateful.pp [hunk 10]
```diff
@@ -68,20 +68,34 @@
...
 
-  iptables_rule { "tcp_${name}":
-    first    => $first,
-    absolute => $absolute,
-    order    => $order,
-    apply_to => $apply_to,
-    content  => template("${module_name}/allow_tcp_udp_services.erb")
+  if $iptables::use_firewalld {
+    iptables::firewalld::rule { "tcp_${name}":
+      trusted_nets => $trusted_nets,
```
Comment (lnemsick-simp) [inline manifests/listen/tcp_stateful.pp]: $first and $absolute do not apply?

File: manifests/listen/udp.pp [hunk 11]
```diff
@@ -68,20 +68,34 @@
...
 
-  iptables_rule { "udp_${name}":
-    first    => $first,
-    absolute => $absolute,
-    order    => $order,
-    apply_to => $apply_to,
-    content  => template("${module_name}/allow_tcp_udp_services.erb")
+  if $iptables::use_firewalld {
+    iptables::firewalld::rule { "udp_${name}":
+      trusted_nets => $trusted_nets,
```
Comment (lnemsick-simp) [inline manifests/listen/udp.pp]: $first and $absolute do not apply?

File: functions/use_firewalld.pp [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+# @param enable
+#   The type of enablement to use
+#
+#   * true      => Do the right thing based on the underlying OS
+#   * false     => Return `false`
+#   * firewalld => Force `firewalld` if available
+#
+# @return [Boolean]
+#
+function iptables::use_firewalld(
```
Comment (lnemsick-simp) [inline functions/use_firewalld.pp]: Are you assuming this is fully tested in the acceptance test?  Didn't see a unit test.

File: manifests/firewalld/rule.pp [hunk 13]
```diff
@@ -0,0 +1,239 @@
+# Add firewall rules via firewalld
```
Comment (lnemsick-simp) [inline manifests/firewalld/rule.pp]: Are you assuming this is fully tested in the spec/defines/listen/firewalld/*_spec.rb tests and the acceptance tests?  Didn't see an explicit unit test for it.

File: manifests/listen/icmp.pp (hunk 9 above)
Comment (trevor-vaughan) [reply manifests/listen/icmp.pp]: @lnemsick-simp No, they do not. Firewalld doesn't have any concept of ordering at this time.

File: manifests/listen/all.pp (hunk 8 above)
Comment (trevor-vaughan) [reply manifests/listen/all.pp]: @lnemsick-simp No, firewalld does not have any concept of specific ordering at this time so it's all roughly "best guess"

File: functions/use_firewalld.pp (hunk 12 above)
Comment (trevor-vaughan) [reply functions/use_firewalld.pp]: The unit tests are through the Hieradata in the 'listen/firewalld' defined type tests.

File: manifests/firewalld/rule.pp (hunk 13 above)
Comment (trevor-vaughan) [reply manifests/firewalld/rule.pp]: It should be in now with the latest commits.

**Comment by trevor-vaughan** [comment]
:hourglass_flowing_sand: https://gitlab.com/simp/pupmod-simp-iptables/pipelines/101039354

File: manifests/rule.pp [hunk 14]
```diff
@@ -76,22 +73,33 @@
...
-    header   => $header,
-    content  => $content,
-    apply_to => $apply_to
+  include iptables
+
+  if $iptables::use_firewalld {
+    $_caller = simplib::caller()
+
+    notify { 'iptables::rule with firewalld':
+      message  => "iptables::rule cannot be used directly in firewalld mode, please use iptables::firewalld::rule => Called from ${_caller}",
```
Comment (lnemsick-simp) [inline manifests/rule.pp]: IMHO CHANGELOG and README.md are confusing.  They say iptables::rule will be auto-mapped, but this fails.  Should the documentation be more explicit and indicate that iptables::listen::xxx have been mapped and users must manually map iptables::rule defines and native types?

**Pull request merged**